File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */

const mostFreqCount = map => {
    let count = 0;
    Object.values(map).forEach(num => {
      count = Math.max(count, num);  
    })
    return count;
}

const characterReplacement = (s, k) => {
    let longest = 0;
    let l = 0;
    let count = {};
    for (let r = 0; r < s.length; r++) {
        count[s[r]] = count[s[r]] + 1 || 1;
        while (r - l + 1 - mostFreqCount(count) > k) {
            count[s[l]] -= 1;
            l++;
        }
        longest = Math.max(longest, r - l + 1);
    }
    return longest;
};

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        left, right = 0, 0
        longest = 0
        count = {}
        most_freq_char = 0
        while right < len(s):
            if s[right] in count:
                count[s[right]] += 1
            else:
                count[s[right]] = 1
            window_length = right - left + 1
            most_freq_char = max(most_freq_char, count[s[right]])
            if window_length - most_freq_char <= k:
                longest = max(longest, window_length)
            if window_length - most_freq_char > k:
                count[s[left]] -= 1
                left += 1
            right += 1
        return longest